### PR TITLE
fix(smart-ngrx): make sure setting a row dirty with virtual arrays updates the arrays without flickering

### PR DESCRIPTION
**Pull request by ruancomelli**
# Issue Number: #592

# Body

Verifies the following:
- Should be able to edit an expanded item in virtual array mode
- using `c` character should not close the editor
- when I edit a field and press enter to save, the field should not blink
- when I set a row dirty that has virtual child array, we need to force the array to re-retrieve the IDs for an index
- when I add a row on one screen, it should show up on another screen
- when I delete a row on one screen, it should be removed on another screen
- After garbage collection, I should be able to go back to a previous tab and it should render

## Release Notes

- **New Features**
  - Added keyboard event handling in the Node Editor component for enhanced interactivity.
  - Added a new public method `refetchIndexes` in the `VirtualArray` class to allow fresh fetching of indexes.

- **Improvements**
  - Improved the `VirtualArray` class with a new `fetchedIndexes` property so we can retrieve by array index position as an additional option for dealing with child rows.

- **Bug Fixes**
  - Added safeguards in the `TreeComponentService` to prevent errors with invalid data ranges.
  
- **Tests**
  - Expanded test coverage for `VirtualArray` and introduced tests for `virtualArrayMap` functionality.

### CODE REVIEW

**Comment by ruancomelli** [comment]
This is a benchmark review for experiment `multiple_mermaid_diagram_types`.
Run ID: `multiple_mermaid_diagram_types/benchmark_2024-09-26T14-38-21_v1-23-0-20-g8a080be57-dirty`.

This pull request was cloned from `https://github.com/DaveMBush/SmartNgRX/pull/622`. _(Note: the URL is not a link to avoid triggering a notification on the original pull request.)_
<details>
<summary>Experiment configuration</summary>

```yaml
review_config:
  # User configuration for the review
  # - benchmark - use the user config from the benchmark reviews
  # - <value> - use the value directly
  user_review_config:
    enable_ai_review: true
    enable_rule_comments: false

    enable_complexity_comments: false
    enable_security_comments: false
    enable_tests_comments: false
    enable_comment_suggestions: false
    enable_functionality_review: false

    enable_pull_request_summary: false
    enable_review_guide: true

    enable_approvals: false

  ai_review_config:
    # The model responses to use for the experiment
    # - benchmark - use the model responses from the benchmark reviews
    # - llm - call the language model to generate responses
    model_responses:
      comments_model: benchmark
      comment_area_model: benchmark
      comment_validation_model: benchmark
      comment_suggestion_model: benchmark
      complexity_model: benchmark
      functionality_model: benchmark
      security_model: benchmark
      tests_model: benchmark
      pull_request_summary_model: benchmark
      review_guide_model: llm
      overall_comments_model: benchmark

# The pull request dataset to run the experiment on
pull_request_dataset:
- https://github.com/sourcery-ai/core/pull/4607
- https://github.com/sourcery-ai/core/pull/4631
- https://github.com/sourcery-ai/core/pull/4647
    # CodeRabbit examples:
- https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
- https://github.com/2lambda123/galaxyproject-galaxy/pull/12
- https://github.com/a0v0/avtoolz/pull/79
- https://github.com/adityask98/Hotaru/pull/10
- https://github.com/agdas/vscode/pull/2
- https://github.com/agluszak/hirschgarten/pull/2
- https://github.com/alexsnow348/insightface/pull/46
- https://github.com/alikuxac/utilities/pull/10
- https://github.com/AlphaDev87/timba-api/pull/49
- https://github.com/AngeloTadeucci/Maple2/pull/239
- https://github.com/AngeloTadeucci/Maple2.File/pull/36
- https://github.com/AngeloTadeucci/Maple2/pull/233
    # Examples where CodeRabbit does not generate diagrams
- https://github.com/baptisteArno/typebot.io/pull/1778
- https://github.com/btipling/foundations/pull/33
- https://github.com/btipling/foundations/pull/31
- https://github.com/chintu-777/jaeger/pull/1
- https://github.com/coji/remix-docs-ja/pull/55
- https://github.com/DaveMBush/SmartNgRX/pull/622
- https://github.com/DaveMBush/SmartNgRX/pull/481
- https://github.com/dkittle/party-connections/pull/6
- https://github.com/Drajad-Kusuma-Adi/onstudy-backend/pull/6
- https://github.com/imaami/libcanth/pull/2
    # Requested by Tim
- https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
- https://github.com/2lambda123/galaxyproject-galaxy/pull/12
- https://github.com/a0v0/avtoolz/pull/79
- https://github.com/adityask98/Hotaru/pull/10
- https://github.com/agdas/vscode/pull/2
- https://github.com/agluszak/hirschgarten/pull/2
- https://github.com/4DNucleome/PartSeg/pull/1183
- https://github.com/ably/ably-cocoa/pull/1973
- https://github.com/ably-labs/ably-chat-swift/pull/54
- https://github.com/ably-labs/ably-chat-swift/pull/35
  # Including database changes to evaluate entity-relationship diagrams
- https://github.com/sourcery-ai/core/pull/4579
  # TODO: find some with Alembic

# Questions to ask to label the review comments
review_comment_labels: []
# - label: correct
#   question: Is this comment correct?

# Benchmark reviews generated by running
#   python -m scripts.experiment benchmark <experiment_name>
benchmark_reviews: []

```

</details>

**Comment by SourceryAI** [comment]
## Reviewer's Guide by Sourcery

This pull request addresses several issues related to virtual arrays in the smart-ngrx library, focusing on improving performance, reducing flickering, and enhancing the overall functionality of the tree component. The changes include modifications to the VirtualArray class, introduction of a new VirtualArrayMap, and updates to various components and services to support these improvements.

### Sequence Diagrams

#### Sequence Diagram: VirtualArray Fetching Process

```mermaid
sequenceDiagram
    participant C as Component
    participant VA as VirtualArray
    participant S as Store
    participant VAM as VirtualArrayMap
    C->>VA: Access index
    VA->>VA: Check fetchedIndexes
    alt Index not fetched
        VA->>S: Dispatch loadByIndexes action
        VA->>VA: Update fetchedIndexes
    end
    VA->>VAM: Update VirtualArrayMap
    VA-->>C: Return value
```

#### Class Diagram: VirtualArray and Related Classes

```mermaid
classDiagram
    class VirtualArray {
        +rawArray: string[]
        +fetchedIndexes: boolean[]
        +length: number
        +refetchIndexes()
        -dispatchLoadByIndexes()
    }
    class VirtualArrayMap {
        -entityMap: Map
        +get(feature, entity, id, field)
        +set(feature, entity, id, field, value)
        +remove(feature, entity, id)
    }
    class ArrayProxy {
        +rawArray: VirtualArray
    }
    VirtualArray --o ArrayProxy
    VirtualArrayMap --o VirtualArray
```

### File-Level Changes

| Change | Details | Files |
| ------ | ------- | ----- |
| Enhance VirtualArray class with new features and optimizations | <ul><li>Add fetchedIndexes property to track which indexes have been fetched</li><li>Implement refetchIndexes method to reset fetchedIndexes</li><li>Optimize dispatchLoadByIndexes to avoid unnecessary dispatches</li><li>Update array access behavior to handle frozen arrays</li></ul> | `libs/smart-ngrx/src/selector/virtual-array.class.ts`<br/>`libs/smart-ngrx/src/selector/virtual-array.class.spec.ts` |
| Introduce VirtualArrayMap for efficient management of virtual arrays | <ul><li>Implement get, set, and remove methods for VirtualArrayMap</li><li>Add tests for VirtualArrayMap functionality</li><li>Integrate VirtualArrayMap with existing components and services</li></ul> | `libs/smart-ngrx/src/selector/virtual-array-map.const.ts`<br/>`libs/smart-ngrx/src/selector/virtual-array-map.const.spec.ts`<br/>`libs/smart-ngrx/src/actions/action.service.ts` |
| Improve tree component performance and stability | <ul><li>Update TreeComponentService to handle invalid data ranges</li><li>Remove debounceTime from location selection in TreeComponent</li><li>Enhance NodeEditorComponent with keyboard event handling</li></ul> | `apps/demo/src/app/shared/components/tree/tree-component.service.ts`<br/>`apps/demo/src/app/routes/tree-standard/tree.component.ts`<br/>`apps/demo/src/app/shared/components/node-editor/node-editor.component.ts`<br/>`apps/demo/src/app/shared/components/node-editor/node-editor.component.html` |
| Refactor and optimize row proxy functionality | <ul><li>Implement forceRefetchOfVirtualIndexes function</li><li>Update rowProxy function to use forceRefetchOfVirtualIndexes</li><li>Add tests for forceRefetchOfVirtualIndexes</li></ul> | `libs/smart-ngrx/src/row-proxy/force-refetch-of-virtual-indexes.function.ts`<br/>`libs/smart-ngrx/src/row-proxy/force-refetch-of-virtual-indexes.function.spec.ts`<br/>`libs/smart-ngrx/src/row-proxy/row-proxy.function.ts` |
| Update and optimize various utility functions and tests | <ul><li>Modify mergeVirtualArrays function to handle undefined indexes</li><li>Update convertChildrenToVirtualArray function to use VirtualArrayMap</li><li>Expand test coverage for various components and functions</li></ul> | `libs/smart-ngrx/src/common/merge-virtual-arrays.function.spec.ts`<br/>`libs/smart-ngrx/src/selector/convert-children-to-virtual-array.function.ts`<br/>`libs/smart-ngrx/src/selector/convert-children-to-virtual-array.function.spec.ts` |

---

<details>
<summary>Tips and commands</summary>

#### Interacting with Sourcery

- **Trigger a new review:** Comment `@sourcery-ai review` on the pull request.
- **Continue discussions:** Reply directly to Sourcery's review comments.
- **Generate a GitHub issue from a review comment:** Ask Sourcery to create an
  issue from a review comment by replying to it.

#### Customizing Your Experience

Access your [dashboard](https://app.sourcery.ai) to:
- Enable or disable review features such as the Sourcery-generated pull request
  summary, the reviewer's guide, and others.
- Change the review language.
- Add, remove or edit custom review instructions.
- Adjust other review settings.

#### Getting Help

- [Contact our support team](mailto:support@sourcery.ai) for questions or feedback.
- Visit our [documentation](https://docs.sourcery.ai) for detailed guides and information.
- Keep in touch with the Sourcery team by following us on [X/Twitter](https://x.com/SourceryAI), [LinkedIn](https://www.linkedin.com/company/sourceryai/) or [GitHub](https://github.com/sourcery-ai).

</details>

File: libs/smart-ngrx/src/actions/action.service.ts [hunk 1]
```diff
@@ -412,5 +412,12 @@ export class ActionService {
         ids: idsToRemove,
       }),
     );
+    // make sure we remove the virtualArray from the map AFTER
+    // we remove the row from the store
+    asapScheduler.schedule(() => {
+      idsToRemove.forEach((id) => {
+        virtualArrayMap.remove(this.feature, this.entity, id);
+      });
+    });
```
Comment (SourceryAI) [inline libs/smart-ngrx/src/actions/action.service.ts:421]: **suggestion (bug_risk):** Consider potential race conditions with asynchronous virtualArrayMap removal

Using asapScheduler to remove entries from virtualArrayMap after dispatching the remove action could potentially lead to race conditions. Consider whether it's possible to remove the entries synchronously after the action dispatch, or if not, implement additional safeguards to ensure consistency between the store and virtualArrayMap.

**Review by SourceryAI — COMMENTED** [review]
Hey @ruancomelli - I've reviewed your changes and they look great!

<details>
<summary>Here's what I looked at during the review</summary>

- 🟡 **General issues**: 1 issue found
- 🟢 **Documentation**: all looks good
</details>

***

[LangSmith trace](https://smith.langchain.com/o/7d8518ad-6dd3-5c7b-96d2-757769296335/projects/p/88560225-5072-48bc-967c-dc54da53f265/r/90cf7ee5-ef4e-429f-bc68-95ef4e8d8593?poll=true)

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment to tell me if it was helpful.
</sub>

**Pull request closed without merging**